tempodb/encoding/v2/data_reader.go

```diff
@@ -23,41 +27,97 @@ var constDataHeader = &dataHeader{}
 
 // NewDataReader constructs a v2 DataReader that handles paged...reading
 func NewDataReader(r backend.ContextReader, encoding backend.Encoding) (common.DataReader, error) {
-	v2DataReader := &dataReader{
-		contextReader: r,
-		dataReader:    v0.NewDataReader(r),
-	}
-
-	// wrap the paged reader in a compressed/v1 reader and return that
-	v1DataReader, err := v1.NewNestedDataReader(v2DataReader, encoding)
+	pool, err := getReaderPool(encoding)
 	if err != nil {
 		return nil, err
 	}
 
-	return v1DataReader, nil
+	return &dataReader{
+		contextReader: r,
+		pool:          pool,
+	}, nil
 }
 
 // Read implements common.DataReader
 func (r *dataReader) Read(ctx context.Context, records []common.Record, pagesBuffer [][]byte, buffer []byte) ([][]byte, []byte, error) {
-	pagesBuffer, buffer, err := r.dataReader.Read(ctx, records, pagesBuffer, buffer)
+	if len(records) == 0 {
+		return nil, buffer, nil
+	}
+
+	start := records[0].Start
+	length := uint32(0)
+	for _, record := range records {
+		length += record.Length
+	}
+
+	if cap(buffer) < int(length) {
+		buffer = make([]byte, length)
+	}
+	buffer = buffer[:length]
+	_, err := r.contextReader.ReadAt(ctx, buffer, int64(start))
 	if err != nil {
 		return nil, nil, err
 	}
 
-	for i := range pagesBuffer {
-		p, err := unmarshalPageFromBytes(pagesBuffer[i], constDataHeader)
+	compressedPagesBuffer := make([][]byte, len(records))
+
+	cursor := uint32(0)
+	previousEnd := uint64(0)
+	for i, record := range records {
+		end := cursor + record.Length
+		if end > uint32(len(buffer)) {
+			return nil, nil, fmt.Errorf("record out of bounds while reading pages: %d, %d, %d, %d", cursor, record.Length, end, len(buffer))
+		}
+
+		if previousEnd != record.Start && previousEnd != 0 {
+			return nil, nil, fmt.Errorf("non-contiguous pages requested from dataReader: %d, %+v", previousEnd, record)
+		}
+
+		compressedPagesBuffer[i] = buffer[cursor:end]
+		cursor += record.Length
+		previousEnd = record.Start + uint64(record.Length)
+	}
+
+	// read and strip page data
+	compressedPages := make([][]byte, 0, len(compressedPagesBuffer))
+	for _, v0Page := range compressedPagesBuffer {
+		page, err := unmarshalPageFromBytes(v0Page, constDataHeader)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		compressedPages = append(compressedPages, page.data)
+	}
+
+	// prepare pagesBuffer
+	if cap(pagesBuffer) < len(compressedPages) {
+		// extend pagesBuffer
+		diff := len(compressedPages) - cap(pagesBuffer)
+		pagesBuffer = append(pagesBuffer[:cap(pagesBuffer)], make([][]byte, diff)...)
```

Comment:
`append()` will copy over all the items from old to new array right? and the resultant pagesBuffer array might have a cap larger than `len(compressedPages)`, worth just creating a new array with `cap(pagesBuffer) = len(compressedPages)` ? The reuse is primarily useful when `cap(pagesBuffer) > len(compressedPages)` I guess

Reply:
> and the resultant pagesBuffer array might have a cap larger than len(compressedPages)
I believe after this line len(pagesBuffer) = cap(pagesBuffer) = len(compressedPages)

> worth just creating a new array with cap(pagesBuffer) = len(compressedPages)
by using whatever currently exists in pagesBuffer we reduce byte slice allocations. after this line the first N entries in pagesBuffer already have allocated byte slices that can be reused in `ReadAllWIthBuffer`. if we recreated the slice we would lose that.